OAK-6766: Convert oak-lucene to OSGi R7 annotations

mbaedke opened the pull request.

mbaedke committed:
OAK-6766: Convert oak-lucene to OSGi R7 annotations

Done.

mbaedke committed:
OAK-6766: Convert oak-lucene to OSGi R7 annotations

Fixed handling of component property attribute.

mbaedke committed:
OAK-6766: Convert oak-lucene to OSGi R7 annotations

Simplified component property handling.

reschke reviewed (commented):
Would be good to get feedback from the Lucene team for this. Maybe the issue of "pirvate" properties can be simplified?

mbaedke commented:
> Would be good to get feedback from the Lucene team for this.

Definitely

> Maybe the issue of "private" properties can be simplified?

Now that I realized that they can be retrieved directly from the ComponentContext without needing a reference to the ConfigurationAdmin service (see last commit), I don't think it can get simpler anymore.

jsedding commented:
> I don't think it can get simpler anymore.

Have you tried adding the private property names to the annotation, but leaving the `@Attribute` meta-annotation off? I have a feeling that could work, and you could then access the values more easily. YMMV, I am not 100% certain.

kwin commented on oak-lucene/src/main/java/org/apache/jackrabbit/oak/plugins/index/lucene/LuceneIndexProviderService.java:
```diff
@@ -368,13 +349,19 @@ public class LuceneIndexProviderService {
-        hybridIndex = PropertiesUtil.toBoolean(config.get(PROP_HYBRID_INDEXING), PROP_DISABLED_DEFAULT);
+                PropertiesUtil.toLong(config.luceneIndexStatsUpdateInterval(),
+                        PROP_LUCENE_INDEX_STATS_UPDATE_INTERVAL_DEFAULT) * 1000); // convert seconds to millis
+        boolean disabled = PropertiesUtil.toBoolean(config.disabled(), PROP_DISABLED_DEFAULT);
```
There is no need for a type conversion here. Also the default is always set according to what is defined in the component property type (https://docs.osgi.org/specification/osgi.cmpn/7.0.0/service.component.html#service.component-component.property.types)

mbaedke replied:
Oh yes, missed that one, thx!

mbaedke commented:
@jsedding Yes, I tried that :)

kwin commented on oak-lucene/src/main/java/org/apache/jackrabbit/oak/plugins/index/lucene/LuceneIndexProviderService.java:
```diff
@@ -104,201 +107,178 @@
 
 @SuppressWarnings("UnusedDeclaration")
-@Component(metatype = true, label = "Apache Jackrabbit Oak LuceneIndexProvider")
+@Component(property = { "enableCopyOnReadSupport:Boolean=true", "enableCopyOnWriteSupport:Boolean=true" } )
```
I would really recommend not mixing component property types with regular properties. Just put the private properties to the Configuration annotation type (but without the metatype annotations on the element). Also the default values only need to be defined once then (and do no longer require a dedicated constant) but can be added as literal on the component property type.

mbaedke replied:
Yes, I did that because otherwise the diff tool complains (the properties will not occur in the component.xml). Meanwhile @cziegele enlightened me. I'll fix it.

kwin commented on oak-lucene/src/main/java/org/apache/jackrabbit/oak/plugins/index/lucene/LuceneIndexProviderService.java:
```diff
@@ -368,13 +349,19 @@ public class LuceneIndexProviderService {
+
     @Activate
-    private void activate(BundleContext bundleContext, Map<String, ?> config) throws IOException {
+    private void activate(BundleContext bundleContext, ComponentContext componentContext, Configuration config) throws IOException {
```
As said above mixing regular properties with component property types is rarely necessary, also instead of injecting the ComponentContext you can rely on a Map to inject component properties if it would be necessary (https://docs.osgi.org/specification/osgi.cmpn/7.0.0/service.component.html#service.component-activation.objects)

mbaedke replied:
Thx

mbaedke committed:
OAK-6766: Convert oak-lucene to OSGi R7 annotations

Removed redundant type conversions.
Replaced component properties with comfiguration type methods.

mbaedke commented:
Addressed @kwin 's suggestions.

mbaedke committed:
Merge branch 'trunk' into issue/oak-6766

mbaedke committed:
OAK-6766: Convert oak-lucene to OSGi R7 annotations

Added missing bind/unbind attribute to reference annotation.

mbaedke committed:
Merge branch 'trunk' into issue/oak-6766
